### PR TITLE
acquiretoken builders API for msal 3.x

### PR DESCRIPTION
**Pull request by MarkZuber**

### CODE REVIEW

File: src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            Parameters.ExtraScopesToConsent = extraScopesToConsent;
+            return (T)this;
+        }
+
+        /// <summary>
+        ///     TODO: replicate the options here that we have in ApplicationBuilder for an AuthorityInfo class?
+        /// </summary>
+        /// <param name="authorityUri"></param>
+        /// <returns></returns>
+        public T WithAuthorityOverride(string authorityUri)
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs]: the question here is if we'd want to have more overrides like WithAzureADAuthorityOverride with clouds / audience?

File: src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        /// <summary>
+        /// </summary>
+        /// <param name="scopes"></param>
+        /// <returns></returns>
+        protected T WithScopes(IEnumerable<string> scopes)
+        {
+            Parameters.Scopes = scopes;
+            return (T)this;
+        }
+
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs]: Would we want to have .WithResource(string resource), which in effect could do a scope of resource/.default?
probably not ?

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            return AcquireTokenOnBehalfOfParameterBuilder.Create(scopes, userAssertion);
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="scopes"></param>
+        /// <param name="account"></param>
+        /// <returns></returns>
+        public static AcquireTokenSilentParameterBuilder CreateSilent(IEnumerable<string> scopes, IAccount account)
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/ApiConfig/AcquireTokenBuilder.cs]: I'm not sure for this one. 
Also AcquireTokenSilent has 2 overrides (one with a force Boolean parameter)

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        {
+            Parameters.ForceRefresh = forceRefresh;
+            return this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="withCertificate"></param>
+        /// <returns></returns>
+        public AcquireTokenForClientParameterBuilder WithCertificate(bool withCertificate)
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: I don't understand this override? would we want to override the certificate set in the CCA builder?

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        internal static AcquireTokenForClientParameterBuilder Create(IEnumerable<string> scopes)
+        {
+            return new AcquireTokenForClientParameterBuilder().WithScopes(scopes);
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="forceRefresh"></param>
+        /// <returns></returns>
+        public AcquireTokenForClientParameterBuilder WithForceRefresh(bool forceRefresh)
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: Good catch as ClientCreds uses the app token cache.

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenOnBehalfOfParameterBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            Parameters.UserAssertion = userAssertion;
+            return this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="withCertificate"></param>
+        /// <returns></returns>
+        public AcquireTokenOnBehalfOfParameterBuilder WithCertificate(bool withCertificate)
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/ApiConfig/AcquireTokenOnBehalfOfParameterBuilder.cs]: Oh! that's the cert roll-up?

File: src/Microsoft.Identity.Client/ApiConfig/GetAuthorizationRequestUrlParameterBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        internal static GetAuthorizationRequestUrlParameterBuilder Create(IEnumerable<string> scopes)
+        {
+            return new GetAuthorizationRequestUrlParameterBuilder().WithScopes(scopes);
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="redirectUri"></param>
+        /// <returns></returns>
+        public GetAuthorizationRequestUrlParameterBuilder WithRedirectUri(string redirectUri)
```
Comment (jmprieur) [inline src/Microsoft.Identity.Client/ApiConfig/GetAuthorizationRequestUrlParameterBuilder.cs]: I don't understand this scenario? we would not use the RedirectUri set in the CCA?

File: src/Microsoft.Identity.Client/ApiConfig/IAcquireTokenForClientParameters.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    /// </summary>
+    public interface IAcquireTokenForClientParameters : IAcquireTokenCommonParameters
+    {
+        /// <summary>
+        /// </summary>
+        bool ForceRefresh { get; }
+
+        /// <summary>
+        /// </summary>
+        bool WithForClientCertificate { get; }
```
Comment (jennyf19) [inline src/Microsoft.Identity.Client/ApiConfig/IAcquireTokenForClientParameters.cs]: >WithFor [](start = 13, length = 7)

this again...seems weird.

File: src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs (hunk 1 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs]: Yes, I think we need to look at this the same way we do authorities in the app builder scenario (thus the TODO in the summary of this method).  I can look at taking what we have finalized in the app builder scenario and putting it here as well.

---
In reply to: [246267784](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/766#discussion_r246267784) [](ancestors = 246267784)

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenBuilder.cs (hunk 3 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ApiConfig/AcquireTokenBuilder.cs]: What part aren't you sure of here?  Also, there's a WithForceRefresh() in the AcquireTokenSilentParameterBuilder to cover this overload scenario.

---
In reply to: [246268023](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/766#discussion_r246268023) [](ancestors = 246268023)

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 4 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: In our current public API, we have:

AcquireTokenForClientAsync(IEnumerable<string> scopes)
AcquireTokenForClientAsync(IEnumerable<string> scopes, bool forceRefresh)
AcquireTokenForClientWithCertificateAsync(IEnumerable<string> scopes)
AcquireTokenForClientWithCertificateAsync(IEnumerable<string> scopes, bool forceRefresh)

The WithCertificate() allows us to differentiate between the AcquireTokenForClient and AcquireTokenForClientWithCertificate versions above.  I'm open to suggestions on how to improve this API surface.  This was put here to ensure we can cover all existing permutations of the public API.


---
In reply to: [246268229](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/766#discussion_r246268229) [](ancestors = 246268229)

File: src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs (hunk 2 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ApiConfig/AbstractAcquireTokenParameterBuilder.cs]: discussed with jm and others, not doing this.

---
In reply to: [246267901](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/766#discussion_r246267901) [](ancestors = 246267901)

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenParameters.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                                            IGetAuthorizationRequestUrlParameters,
+                                            IAcquireTokenSilentParameters
+    {
+        public string AuthorizationCode { get; internal set; }
+        public bool ForceRefresh { get; internal set; }
+        public bool WithForClientCertificate { get; internal set; }
+
+        // Interactive Parameters
+        public bool UseEmbeddedWebView { get; internal set; }
+        public UIBehavior UiBehavior { get; internal set; }
```
[the remark this answers is not in the document]
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ApiConfig/AcquireTokenParameters.cs]: That's currently a public api breaking change without a great migration story other than a compile break.  I'm going to hold for now.

---
In reply to: [246491034](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/766#discussion_r246491034) [](ancestors = 246491034)

File: src/Microsoft.Identity.Client/ApiConfig/GetAuthorizationRequestUrlParameterBuilder.cs (hunk 7 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ApiConfig/GetAuthorizationRequestUrlParameterBuilder.cs]: ConfidentialClientApplication.cs today has this method:

        public async Task<Uri> GetAuthorizationRequestUrlAsync(IEnumerable<string> scopes, string redirectUri, string loginHint,
            string extraQueryParameters, IEnumerable<string> extraScopesToConsent, string authority)

It takes a redirectUri so I have that ability here.  Should we not have it for this case?

---
In reply to: [246268634](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/766#discussion_r246268634) [](ancestors = 246268634)

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 4 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: Updated this to be WithSendX5C as we discussed today.

---
In reply to: [246501111](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/766#discussion_r246501111) [](ancestors = 246501111,246268229)

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenBuilder.cs (hunk 3 above)
Comment (MarkZuber) [reply src/Microsoft.Identity.Client/ApiConfig/AcquireTokenBuilder.cs]: I think from our discussion we're resolved.  Please reactivate if there are further issues.

---
In reply to: [246500144](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/766#discussion_r246500144) [](ancestors = 246500144,246268023)

**Comment by MarkZuber** [comment]
Moving this code to merge into dev3x instead of master

**Pull request closed without merging**